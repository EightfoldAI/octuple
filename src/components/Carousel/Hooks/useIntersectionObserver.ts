import { useCallback, useLayoutEffect, useRef, useState } from 'react';
import type { Item, Refs, visibleElements } from '../Carousel.types';
import { getNodesFromRefs, observerEntriesToItems } from '../Utilities';
import { observerOptions } from '../Settings';
import ItemsMap from '../ItemsMap';
import { requestAnimationFrameWrapper } from '../../../shared/utilities';

interface IntersectionProps {
  items: ItemsMap;
  itemsChanged: string;
  options: typeof observerOptions;
  refs: Refs;
}

export const useIntersectionObserver = ({
  items,
  itemsChanged,
  refs,
  options,
}: IntersectionProps) => {
  const observer: { current?: IntersectionObserver } = useRef();

  const [visibleElementsWithSeparators, setVisibleElementsWithSeparators] =
    useState<visibleElements>([]);

  const throttleTimer: { current: number } = useRef(
    +setTimeout(() => void 0, 0)
  );

  const intersectionObserverCallback = useCallback(
    (entries: IntersectionObserverEntry[]) => {
      items.set(observerEntriesToItems(entries, options));

      clearTimeout(throttleTimer.current);
      throttleTimer.current = +setTimeout(() => {
        requestAnimationFrameWrapper(() => {
          setVisibleElementsWithSeparators((currentVisible) => {
            const newVisibleElements = items
              .getVisible()
              .map((el: Item) => el[1].key);
            if (
              JSON.stringify(currentVisible) !==
              JSON.stringify(newVisibleElements)
            ) {
              return newVisibleElements;
            }
            return currentVisible;
          });
        });
      }, options.throttle);
    },
    [items, options]
  );

  useLayoutEffect(() => {
    const elements: HTMLElement[] = getNodesFromRefs(refs);
    const observerInstance: IntersectionObserver =
      observer.current ||
      new IntersectionObserver(intersectionObserverCallback, options);
    observer.current = observerInstance;
    elements.forEach((elem: Element) => observerInstance.observe(elem));

    return () => {
      clearTimeout(throttleTimer.current);
      observerInstance.disconnect();
      observer.current = undefined;
    };
  }, [intersectionObserverCallback, itemsChanged, options, refs]);

  return {
    visibleElementsWithSeparators,
  };
};
