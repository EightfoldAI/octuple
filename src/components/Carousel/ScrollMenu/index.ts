import { ScrollMenu, VisibilityContext } from './ScrollMenu';

export { ScrollMenu, VisibilityContext };
