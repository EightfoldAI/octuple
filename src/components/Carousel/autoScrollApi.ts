'use client';

import {
  filterSeparators,
  scrollToItem,
  scrollBySingleItem,
  getItemElementById,
  getItemElementByIndex,
} from './Utilities';
import ItemsMap from './ItemsMap';
import type {
  CustomScrollBehavior,
  IntersectionObserverItem,
  ItemOrElement,
  scrollToItemOptions,
  visibleElements,
} from './Carousel.types';

export const autoScrollApi = (
  items: ItemsMap,
  visibleElementsWithSeparators: visibleElements = [],
  boundaryElement?: React.MutableRefObject<HTMLElement | null>,
  transitionOptions?: {
    duration?: number;
    ease?: (t: number) => number;
    behavior: string | Function;
  },
  rtl?: boolean,
  noPolyfill?: boolean
) => {
  const visibleElements: visibleElements = filterSeparators(
    visibleElementsWithSeparators
  );

  const isFirstItemVisible: boolean = !!items.first()?.visible;
  const isLastItemVisible: boolean = !!items.last()?.visible;

  const getItemById: (id: string) => IntersectionObserverItem = (id: string) =>
    items.find((value) => value[1].key === String(id))?.[1];

  const getItemByIndex: (index: number | string) => IntersectionObserverItem = (
    index: number | string
  ) => items.find((el) => String(el[1].index) === String(index))?.[1];

  const isItemVisible: (id: string) => boolean = (id: string) =>
    visibleElements.includes(String(id));

  const getPrevItemGroup: () => IntersectionObserverItem = () =>
    items.prevGroup(items.getVisible()?.[0]?.[1]);

  const getPrevElementGroup: () => IntersectionObserverItem = () =>
    items.prevGroup(items.getVisibleElements()?.[0]?.[1], true);

  const getNextItemGroup: () => IntersectionObserverItem = () =>
    items.nextGroup(items.getVisible()?.slice?.(-1)?.[0]?.[1]);

  const getNextElementGroup: () => IntersectionObserverItem = () =>
    items.nextGroup(items.getVisibleElements()?.slice?.(-1)?.[0]?.[1], true);

  const getPrevItem: () => IntersectionObserverItem = () =>
    items.prev(items.getVisible()?.[0]?.[1]);

  const getPrevElement: () => IntersectionObserverItem = () =>
    items.prev(items.getVisibleElements()?.[0]?.[1], true);

  const getNextItem: () => IntersectionObserverItem = () =>
    items.next(items.getVisible()?.slice?.(-1)?.[0]?.[1]);

  const getNextElement: () => IntersectionObserverItem = () =>
    items.next(items.getVisibleElements()?.slice?.(-1)?.[0]?.[1], true);

  const isLastItem: (id: string) => boolean = (id: string) =>
    items.last() === getItemById(id);

  const scrollPrevGroup = <T>(
    behavior?: CustomScrollBehavior<T>,
    inline?: ScrollLogicalPosition,
    block?: ScrollLogicalPosition,
    {
      duration,
      ease,
      boundary = boundaryElement?.current,
    }: scrollToItemOptions = {}
  ) => {
    const _behavior = (behavior ??
      transitionOptions?.behavior) as ScrollBehavior;

    return scrollToItem(
      getPrevItemGroup(),
      _behavior,
      inline || 'end',
      block || 'nearest',
      {
        boundary,
        duration: duration ?? transitionOptions?.duration,
        ease: ease ?? transitionOptions?.ease,
      },
      rtl || noPolyfill
    );
  };

  const scrollNextGroup = <T>(
    behavior?: CustomScrollBehavior<T>,
    inline?: ScrollLogicalPosition,
    block?: ScrollLogicalPosition,
    {
      duration,
      ease,
      boundary = boundaryElement?.current,
    }: scrollToItemOptions = {}
  ) => {
    const _behavior = (behavior ??
      transitionOptions?.behavior) as ScrollBehavior;

    return scrollToItem(
      getNextItemGroup(),
      _behavior,
      inline || 'start',
      block || 'nearest',
      {
        boundary,
        duration: duration ?? transitionOptions?.duration,
        ease: ease ?? transitionOptions?.ease,
      },
      rtl || noPolyfill
    );
  };

  const scrollPrev = <T>(
    behavior?: CustomScrollBehavior<T>,
    inline?: ScrollLogicalPosition,
    block?: ScrollLogicalPosition,
    {
      duration,
      ease,
      boundary = boundaryElement?.current,
    }: scrollToItemOptions = {}
  ) => {
    const _behavior = (behavior ??
      transitionOptions?.behavior) as ScrollBehavior;

    return scrollToItem(
      getPrevItem(),
      _behavior,
      inline || 'end',
      block || 'nearest',
      {
        boundary,
        duration: duration ?? transitionOptions?.duration,
        ease: ease ?? transitionOptions?.ease,
      },
      rtl || noPolyfill
    );
  };

  const scrollNext = <T>(
    behavior?: CustomScrollBehavior<T>,
    inline?: ScrollLogicalPosition,
    block?: ScrollLogicalPosition,
    {
      duration,
      ease,
      boundary = boundaryElement?.current,
    }: scrollToItemOptions = {}
  ) => {
    const _behavior = (behavior ??
      transitionOptions?.behavior) as ScrollBehavior;

    return scrollToItem(
      getNextItem(),
      _behavior,
      inline || 'start',
      block || 'nearest',
      {
        boundary,
        duration: duration ?? transitionOptions?.duration,
        ease: ease ?? transitionOptions?.ease,
      },
      rtl || noPolyfill
    );
  };

  return {
    getItemById,
    getItemElementById,
    getItemByIndex,
    getItemElementByIndex,
    getNextItem,
    getNextItemGroup,
    getNextElement,
    getNextElementGroup,
    getPrevItem,
    getPrevItemGroup,
    getPrevElement,
    getPrevElementGroup,
    isFirstItemVisible,
    isItemVisible,
    isLastItem,
    isLastItemVisible,
    scrollBySingleItem: <T>(
      target?: ItemOrElement,
      behavior?: CustomScrollBehavior<T>,
      direction?: string,
      gap?: number,
      offset?: number
    ) => {
      const _behavior: string | Function =
        behavior ?? transitionOptions?.behavior;

      return scrollBySingleItem(
        target,
        _behavior as ScrollBehavior | CustomScrollBehavior<T>,
        direction,
        gap,
        offset
      );
    },
    scrollNext,
    scrollNextGroup,
    scrollPrev,
    scrollPrevGroup,
    scrollToItem: <T>(
      target?: ItemOrElement,
      behavior?: CustomScrollBehavior<T>,
      inline?: ScrollLogicalPosition,
      block?: ScrollLogicalPosition,
      options?: scrollToItemOptions
    ) => {
      const _behavior: string | Function =
        behavior ?? transitionOptions?.behavior;

      return scrollToItem(
        target,
        _behavior as ScrollBehavior | CustomScrollBehavior<T>,
        inline,
        block,
        {
          boundary: boundaryElement?.current,
          ...options,
          duration: options?.duration ?? transitionOptions?.duration,
          ease: options?.ease ?? transitionOptions?.ease,
        }
      );
    },
    visibleElements,
    visibleElementsWithSeparators,

    visibleItems: visibleElementsWithSeparators,
    visibleItemsWithoutSeparators: visibleElements,
  };
};

export interface autoScrollApiType extends ReturnType<typeof autoScrollApi> {
  initComplete: boolean;
  items: ItemsMap;
  scrollContainer: React.RefObject<HTMLElement | null>;
  visibleElements: visibleElements;
  visibleElementsWithSeparators: visibleElements;
}
