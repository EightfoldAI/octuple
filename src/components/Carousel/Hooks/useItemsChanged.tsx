import React, { useMemo, useState, useEffect } from 'react';
import { ItemType, separatorString, visibleElements } from '../Carousel.types';
import ItemsMap from '../ItemsMap';
import { getItemId } from '../Utilities';

const getItemsIdFromChildren = (
    children: ItemType | ItemType[] | undefined
): string[] => React.Children.toArray(children).map(getItemId).filter(Boolean);

export const useItemsChanged = (
    menuItems: ItemType | ItemType[] | undefined,
    items: ItemsMap
): string => {
    const [hash, setHash] = useState<string>('');

    const domNodes: string[] = useMemo(
        () => getItemsIdFromChildren(menuItems),
        [menuItems]
    );

    useEffect(() => {
        const hash: string = domNodes.filter(Boolean).join('');
        const allItems: visibleElements = items.toItemsWithoutSeparators();
        const removed: string[] = allItems.filter(
            (item) => !domNodes.includes(item)
        );
        removed.forEach((item: string) => {
            const isLast: boolean = items.last()?.key === item;
            const lastSeparator: string =
                (isLast && items.prev(item)?.key) || '';

            items.delete(lastSeparator);
            items.delete(`${item}${separatorString}`);
            items.delete(item);
        });

        setHash(hash);
    }, [domNodes, items]);

    return hash;
};
