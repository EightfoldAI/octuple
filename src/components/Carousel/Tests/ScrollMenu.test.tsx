import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { ScrollMenu, VisibilityContext } from '../ScrollMenu';
import {
  itemClassName,
  ItemType,
  ScrollMenuProps,
  separatorClassName,
} from '../Carousel.types';
import { autoScrollApiType } from '../autoScrollApi';
import { useIntersectionObserver } from '../Hooks/useIntersectionObserver';
import { act, fireEvent, render } from '@testing-library/react';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

jest.mock('../Hooks/useIntersectionObserver');

const scrollContainerClassName: string = 'carousel-auto-scroll-container';
const wrapperClassName: string = 'carousel-auto-scroll';
const defaultItems = ['test1', 'test2'];
// eslint-disable-next-line radar/no-duplicate-string
const defaultItemsWithSeparators = ['test1', 'item1-separator', 'test2'];
const scrollContainerClassNames = 'scroll-class';
const getContext = (context: autoScrollApiType) => {
  const { items: _1, scrollContainer: _2, ...rest } = context;
  return rest;
};
function Child({ itemId }: { itemId: string }) {
  const context = React.useContext(VisibilityContext);

  const text = getContext(context);
  return (
    <div data-testid={itemId} key={itemId}>
      {JSON.stringify({ itemId, ...text })}
    </div>
  );
}
const defaultChildren = defaultItems.map((itemId) => (
  <Child key={itemId} itemId={itemId} />
));

const PreviousButton = () => {
  const context = React.useContext(VisibilityContext);
  return (
    <div className="left-arrow" data-testid="left-arrow">
      {JSON.stringify(getContext(context))}
    </div>
  );
};
const NextButton = () => {
  const context = React.useContext(VisibilityContext);
  return (
    <div className="right-arrow" data-testid="right-arrow">
      {JSON.stringify(getContext(context))}
    </div>
  );
};

interface SetupProps extends Omit<ScrollMenuProps, 'children'> {
  rerender?: Function;
  children?: ItemType | ItemType[];
}
const setup = ({
  children = defaultChildren,
  rerender,
  ...props
}: SetupProps = {}) => {
  if (rerender) {
    rerender(<ScrollMenu {...props}>{children}</ScrollMenu>);
    return render(<div />);
  }
  return render(<ScrollMenu {...props}>{children}</ScrollMenu>);
};

const options = {
  ratio: 0.9,
  root: null as any,
  rootMargin: '8px',
  threshold: [0, 0.25, 0.5, 0.75, 1],
  throttle: 100,
};

describe('ScrollMenu', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });
  beforeEach(() => {
    jest.resetAllMocks();
    jest.resetAllMocks();
  });
  afterEach(() => {
    matchMedia.clear();
  });

  test('should render without props', () => {
    (useIntersectionObserver as jest.Mock).mockReturnValue({
      visibleElementsWithSeparators: defaultItemsWithSeparators,
    });
    const { container } = setup();

    expect(container.firstChild).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  describe('', () => {
    test('should pass props to it', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const { container } = setup();

      expect(container.firstChild).toBeTruthy();
      expect(useIntersectionObserver).toHaveBeenCalled();
      const call = (useIntersectionObserver as jest.Mock).mock.calls[0].slice(
        -1
      )[0];
      expect(call).toMatchObject({
        items: expect.any(Object),
        itemsChanged: '',
        options,
        refs: Object.fromEntries(
          defaultItems.map((_, ind: number) => [
            String(ind),
            { current: expect.any(Element) },
          ])
        ),
      });
    });
  });

  describe('onInit', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    // eslint-disable-next-line radar/no-duplicate-string
    test('should fire with publicApi', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onInit = jest.fn();
      const { container } = setup({ onInit });

      expect(container.firstChild).toBeTruthy();

      expect(onInit).toHaveBeenCalledTimes(1);
      const call = onInit.mock.calls[0][0];
      comparePublicApi(call);
    });

    test.skip('should return initComplete false on first render(when visibleElementsWithSeparators empty)', () => {
      (useIntersectionObserver as jest.Mock)
        .mockReturnValueOnce({
          visibleElementsWithSeparators: [],
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: [],
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: defaultItemsWithSeparators,
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: defaultItemsWithSeparators,
        });
      const onInit = jest.fn();

      const { container, rerender } = setup({ onInit });

      expect(onInit).not.toHaveBeenCalled();
      const textContent1 = container.firstChild!.textContent;
      expect(textContent1!.includes('"initComplete":false')).toBeTruthy();

      setup({ onInit, rerender });

      expect(onInit).toHaveBeenCalledTimes(1);
      const call = onInit.mock.calls[0][0];
      expect(call.initComplete).toEqual(true);
      const textContent2 = container.firstChild!.textContent;
      expect(textContent2!.includes('"initComplete":true')).toBeTruthy();
    });
  });

  describe('apiRef', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test('should fire with publicApi', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const apiRef = {
        current: {},
      } as React.MutableRefObject<autoScrollApiType>;

      const { container } = setup({ apiRef });

      expect(container.firstChild).toBeTruthy();

      comparePublicApi(apiRef.current);
    });
  });

  describe('onUpdate', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test('should fire with publicApi', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onInit = jest.fn();
      const onUpdate = jest.fn();

      const { container } = setup({ onInit, onUpdate });

      expect(container.firstChild).toBeTruthy();

      expect(onInit).toHaveBeenCalledTimes(1);
      expect(onUpdate).toHaveBeenCalledTimes(1);
      const call = onUpdate.mock.calls[0][0];
      comparePublicApi(call);
    });

    test.skip('should not fire if init not complete (when visibleElementsWithSeparators empty)', () => {
      (useIntersectionObserver as jest.Mock)
        .mockReturnValueOnce({
          visibleElementsWithSeparators: [],
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: [],
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: defaultItemsWithSeparators,
        })
        .mockReturnValueOnce({
          visibleElementsWithSeparators: defaultItemsWithSeparators,
        });
      const onInit = jest.fn();
      const onUpdate = jest.fn();

      const { container, rerender } = setup({ onInit, onUpdate });

      expect(onInit).not.toHaveBeenCalled();
      expect(onUpdate).not.toHaveBeenCalled();
      const textContent1 = container.firstChild!.textContent;
      expect(textContent1!.includes('"initComplete":false')).toBeTruthy();

      setup({ onInit, onUpdate, rerender });

      expect(onInit).toHaveBeenCalledTimes(1);
      expect(onUpdate).toHaveBeenCalledTimes(1);
      const call = onUpdate.mock.calls[0][0];
      comparePublicApi(call);
      expect(call.initComplete).toEqual(true);
      const textContent2 = container.firstChild!.textContent;
      expect(textContent2!.includes('"initComplete":true')).toBeTruthy();
    });
  });

  describe('Children and buttons', () => {
    test('LeftArrow, ScrollContainer, MenuItems, RightArrow', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });

      const { container, getByTestId } = setup({
        previousButton: PreviousButton,
        nextButton: NextButton,
        scrollContainerClassNames,
      });

      const LeftArrow = getByTestId('left-arrow');
      const RightArrow = getByTestId('right-arrow');
      const ScrollContainer = container.getElementsByClassName(
        scrollContainerClassName
      )?.[0];

      const context = {
        initComplete: true,
        isFirstItemVisible: false,
        isLastItemVisible: false,
        visibleElementsWithSeparators: ['test1', 'item1-separator', 'test2'],
        visibleItems: ['test1', 'item1-separator', 'test2'],
        visibleElements: ['test1', 'test2'],
        visibleItemsWithoutSeparators: ['test1', 'test2'],
      };

      expect(ScrollContainer.classList.contains(scrollContainerClassName)).toBe(
        true
      );
      expect(LeftArrow.classList.contains('left-arrow')).toBe(true);
      expect(JSON.parse(LeftArrow.textContent!)).toEqual(context);
      expect(RightArrow.classList.contains('right-arrow')).toBe(true);
      expect(JSON.parse(RightArrow.textContent!)).toEqual(context);

      const MenuItems = ScrollContainer.firstChild!;

      defaultItems.forEach((item) =>
        expect(MenuItems.textContent).toContain(item)
      );

      expect(container).toMatchSnapshot();
    });

    test('Hides LeftArrow, RightArrow', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });

      const { container } = setup({
        previousButton: PreviousButton,
        nextButton: NextButton,
        controls: false,
      });

      expect(container.querySelector('.left-arrow')).toBeFalsy();
      expect(container.querySelector('.right-arrow')).toBeFalsy();

      expect(container).toMatchSnapshot();
    });

    test('overlayControls is false', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });

      const { container } = setup({
        previousButton: PreviousButton,
        nextButton: NextButton,
        overlayControls: false,
      });

      expect(container.querySelector('.carousel-next-mask')).toBeFalsy();
      expect(container.querySelector('.carousel-previous-mask')).toBeFalsy();

      expect(container).toMatchSnapshot();
    });
  });

  describe('Events', () => {
    test.skip('should fire onScroll', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onScroll = jest.fn();
      const { container } = setup({ onScroll });

      act(() => {
        fireEvent.scroll(container.firstChild as Element);
      });

      expect(onScroll).toHaveBeenCalledTimes(1);
      const call = onScroll.mock.calls[0][0];
      comparePublicApi(call);
    });

    test('should fire onWheel', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onWheel = jest.fn();
      const { container } = setup({ onWheel });

      act(() => {
        fireEvent.wheel(container.firstChild as Element);
      });

      expect(onWheel).toHaveBeenCalledTimes(1);
      const call = onWheel.mock.calls[0][0];
      comparePublicApi(call);
    });

    test('should fire onMouseDown', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onMouseDown = jest.fn();
      const { container } = setup({ onMouseDown });

      act(() => {
        fireEvent.mouseDown(container.firstChild as Element);
      });

      expect(onMouseDown).toHaveBeenCalled();
      const call = onMouseDown.mock.calls[0][0];
      comparePublicApi(call);
    });

    test('should fire onMouseUp', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onMouseUp = jest.fn();
      const { container } = setup({ onMouseUp });

      act(() => {
        fireEvent.mouseUp(container.firstChild as Element);
      });

      expect(onMouseUp).toHaveBeenCalled();
      const call = onMouseUp.mock.calls[0][0];
      comparePublicApi(call);
    });

    test('should fire onMouseMove', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });
      const onMouseMove = jest.fn();
      const { container } = setup({ onMouseMove });

      act(() => {
        fireEvent.mouseMove(container.firstChild as Element);
      });

      expect(onMouseMove).toHaveBeenCalled();
      const call = onMouseMove.mock.calls[0][0];
      comparePublicApi(call);
    });

    describe('classNames', () => {
      test('should pass classNames', () => {
        (useIntersectionObserver as jest.Mock).mockReturnValue({
          visibleElementsWithSeparators: defaultItemsWithSeparators,
        });

        const itemClassNames = 'item-class';
        const separatorClassNames = 'sep-class';
        const scrollWrapperClassNames = 'wrapper-class';

        const { container } = setup({
          itemClassNames,
          separatorClassNames,
          scrollContainerClassNames,
          scrollWrapperClassNames,
        });

        const Wrapper = container.getElementsByClassName(wrapperClassName)?.[0];

        const ScrollContainer = container.getElementsByClassName(
          scrollContainerClassName
        )?.[0];

        expect(Wrapper.getAttribute('class')).toEqual(
          `${wrapperClassName} ${scrollWrapperClassNames}`
        );

        expect(ScrollContainer.getAttribute('class')).toEqual(
          `${scrollContainerClassName} ${scrollContainerClassNames}`
        );

        const item = ScrollContainer.firstChild as HTMLElement;
        expect(item.getAttribute('class')).toEqual(
          `${itemClassName} ${itemClassNames}`
        );

        const separator = ScrollContainer.childNodes[1] as HTMLElement;
        expect(separator.getAttribute('class')).toEqual(
          `${separatorClassName} ${separatorClassNames}`
        );
      });
    });
  });

  describe('rtl', () => {
    test('should add rtl class', () => {
      (useIntersectionObserver as jest.Mock).mockReturnValue({
        visibleElementsWithSeparators: defaultItemsWithSeparators,
      });

      const itemClassNames = 'item-class';
      const separatorClassNames = 'sep-class';
      const scrollWrapperClassNames = 'wrapper-class';

      const { container } = setup({
        itemClassNames,
        separatorClassNames,
        scrollContainerClassNames,
        scrollWrapperClassNames,
        rtl: true,
      });

      const Wrapper = container.getElementsByClassName(wrapperClassName)?.[0];

      const ScrollContainer = container.getElementsByClassName(
        scrollContainerClassName
      )?.[0];

      expect(Wrapper.getAttribute('class')).toEqual(
        `${wrapperClassName} ${scrollWrapperClassNames}`
      );

      expect(ScrollContainer.getAttribute('class')).toEqual(
        `${scrollContainerClassName} carousel-auto-scroll-container-rtl ${scrollContainerClassNames}`
      );
    });
  });
});

function comparePublicApi(call: autoScrollApiType) {
  const {
    getItemById,
    getItemByIndex,
    getNextElement,
    getNextElementGroup,
    getNextItem,
    getNextItemGroup,
    getPrevElement,
    getPrevElementGroup,
    getPrevItem,
    getPrevItemGroup,
    isItemVisible,
    isLastItem,
    scrollNext,
    scrollNextGroup,
    scrollPrev,
    scrollPrevGroup,
    scrollToItem,
    scrollBySingleItem,
    visibleElementsWithSeparators,
    visibleElements,
    initComplete,
    isFirstItemVisible,
    isLastItemVisible,
    items,
    scrollContainer,
  } = call;

  expect(getItemById).toEqual(expect.any(Function));
  expect(getItemByIndex).toEqual(expect.any(Function));
  expect(getNextElement).toEqual(expect.any(Function));
  expect(getNextElementGroup).toEqual(expect.any(Function));
  expect(getNextItem).toEqual(expect.any(Function));
  expect(getNextItemGroup).toEqual(expect.any(Function));
  expect(getPrevElement).toEqual(expect.any(Function));
  expect(getPrevElementGroup).toEqual(expect.any(Function));
  expect(getPrevItem).toEqual(expect.any(Function));
  expect(getPrevItemGroup).toEqual(expect.any(Function));
  expect(isItemVisible).toEqual(expect.any(Function));
  expect(isLastItem).toEqual(expect.any(Function));
  expect(scrollBySingleItem).toEqual(expect.any(Function));
  expect(scrollNext).toEqual(expect.any(Function));
  expect(scrollNextGroup).toEqual(expect.any(Function));
  expect(scrollPrevGroup).toEqual(expect.any(Function));
  expect(scrollPrev).toEqual(expect.any(Function));
  expect(scrollToItem).toEqual(expect.any(Function));
  expect(visibleElementsWithSeparators).toEqual(defaultItemsWithSeparators);
  expect(visibleElements).toEqual(defaultItems);
  expect(initComplete).toEqual(expect.any(Boolean));
  expect(isFirstItemVisible).toEqual(expect.any(Boolean));
  expect(isLastItemVisible).toEqual(expect.any(Boolean));
  expect(items).toEqual(expect.any(Object));
  expect(scrollContainer).toEqual({ current: expect.any(Element) });
}
