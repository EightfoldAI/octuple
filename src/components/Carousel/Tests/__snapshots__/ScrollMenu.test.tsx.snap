// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollMenu Children and buttons Hides LeftArrow, RightArrow 1`] = `
<div>
  <div
    class="carousel-auto-scroll"
  >
    <div
      class="carousel-auto-scroll-inner-wrapper"
    >
      <ul
        class="carousel-auto-scroll-container"
        style="margin: 0px; padding: 0px;"
      >
        <li
          class="carousel-scroll-menu-item"
          data-index="0"
          data-key="test1"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
        <div
          aria-hidden="true"
          class="carousel-scroll-menu-separator"
          data-index="0.1"
          data-key="test1-separator"
          style="margin-left: 0px; margin-right: 4px;"
        />
        <li
          class="carousel-scroll-menu-item"
          data-index="1"
          data-key="test2"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`ScrollMenu Children and buttons LeftArrow, ScrollContainer, MenuItems, RightArrow 1`] = `
<div>
  <div
    class="carousel-auto-scroll"
  >
    <div
      class="carousel-auto-scroll-inner-wrapper"
    >
      <div
        class="left-arrow"
        data-testid="left-arrow"
      >
        {"isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
      </div>
      <ul
        class="carousel-auto-scroll-container scroll-class"
        style="margin: 0px; padding: 0px;"
      >
        <li
          class="carousel-scroll-menu-item"
          data-index="0"
          data-key="test1"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
        <div
          aria-hidden="true"
          class="carousel-scroll-menu-separator"
          data-index="0.1"
          data-key="test1-separator"
          style="margin-left: 0px; margin-right: 4px;"
        />
        <li
          class="carousel-scroll-menu-item"
          data-index="1"
          data-key="test2"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
      </ul>
      <div
        class="right-arrow"
        data-testid="right-arrow"
      >
        {"isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
      </div>
    </div>
  </div>
</div>
`;

exports[`ScrollMenu should render without props 1`] = `
<div>
  <div
    class="carousel-auto-scroll"
  >
    <div
      class="carousel-auto-scroll-inner-wrapper"
    >
      <ul
        class="carousel-auto-scroll-container"
        style="margin: 0px; padding: 0px;"
      >
        <li
          class="carousel-scroll-menu-item"
          data-index="0"
          data-key="test1"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test1"
          >
            {"itemId":"test1","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
        <div
          aria-hidden="true"
          class="carousel-scroll-menu-separator"
          data-index="0.1"
          data-key="test1-separator"
          style="margin-left: 0px; margin-right: 4px;"
        />
        <li
          class="carousel-scroll-menu-item"
          data-index="1"
          data-key="test2"
          style="margin-left: 0px; margin-right: 4px;"
        >
          <div
            data-testid="test2"
          >
            {"itemId":"test2","isFirstItemVisible":false,"isLastItemVisible":false,"visibleElements":["test1","test2"],"visibleElementsWithSeparators":["test1","item1-separator","test2"],"visibleItems":["test1","item1-separator","test2"],"visibleItemsWithoutSeparators":["test1","test2"],"initComplete":true}
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
